server/src/main/java/org/elasticsearch/common/settings/Settings.java

```diff
@@ -120,10 +119,37 @@ private Settings(Map<String, Object> settings, SecureSettings secureSettings) {
             }
             tree.put(internKeyOrValue(settingEntry.getKey()), internedValue);
         }
-        this.settings = Collections.unmodifiableNavigableMap(tree);
+        return new Settings(Collections.unmodifiableNavigableMap(tree), secureSettings);
+    }
+
+    private Settings(NavigableMap<String, Object> settings, @Nullable SecureSettings secureSettings) {
+        this.settings = settings;
         this.secureSettings = secureSettings;
     }
 
+    /**
+     * Create a merged instance that contains all settings from this instance and the given settings, with the given {@code other} settings
+     * taking precedence. This method is functionally equivalent to using {@link Builder#put(Settings)} twice to merge two settings
+     * instances but much faster.
+     *
+     * @param other settings to merge with this instance's settings.
+     * @return merge settings
+     */
+    public Settings merge(Settings other) {
```

Comment:
Can we please add a unit test for this method? In particular, I think we should ensure there's a test for the precedence behaviour of 'other' over 'this'.